### PR TITLE
Assignments week 2 Ilia Bubnov

### PR DESCRIPTION
**Pull request by MBreathe**

### CODE REVIEW

File: assignments/hackyourtemperature/app.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+app.set("views", "./views");
+
+app.use(express.json());
+
+app.get("/", (req, res) => {
+  //res.render("index");
+  res.send("hello from backend to frontend!");
+});
+app.post("/weather", (req, res) => {
+  getWeather(req, res);
```
Comment (EdwardSmart98) [inline assignments/hackyourtemperature/app.js:18]: This is really good seperation of logic, placing functions/controllers in a different class/method. Good Job!

File: assignments/hackyourtemperature/functions.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    throw response;
+  }
+  return response.json();
+}
+
+export async function getWeather(req, res) {
+  const message = { weatherText: "" };
+  if (!req.body.cityName) {
+    message.weatherText = "You have to fill in city name";
+    return res.status(400).json(message);
```
Comment (EdwardSmart98) [inline assignments/hackyourtemperature/functions.js:15]: We should use a different model for errors than we do for valid responses, as the client may not check the status code, and just if their is a weatherText value. A good practise is all errors get returned in an error object like : 
{
  "errorCode" : "INVALID_REQUEST",
  "errorMessage" : "You have to fill in a city name" 
}

This allows you to build standard error wrappers/handlers in both the back-end and front-end.

File: assignments/hackyourtemperature/functions.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    if (e.status >= 400 && e.status < 500) {
+      message.weatherText =
+        "Error occurred while fetching data, please check your request";
+      return res.status(e.status).json(message);
+    }
+    if (e.status >= 500) {
+      message.weatherText =
+        "Something went wrong on server side, try again later";
+      return res.status(e.status).json(message);
+    }
```
Comment (EdwardSmart98) [inline assignments/hackyourtemperature/functions.js:39]: If the errors status is not set (e.g. the error was thrown from something other that the API request, then this API will not respond properly, good to write at the bottom a single "unknown error".  response.

File: assignments/hackyourtemperature/__tests__/app.test.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import { app } from "../app.js";
+import supertest from "supertest";
+
+const request = supertest(app);
+
+describe("POST /weather", () => {
+  it("Should return an error when cityName is not passed", async () => {
```
Comment (EdwardSmart98) [inline assignments/hackyourtemperature/__tests__/app.test.js:7]: Good checking of edge cases.

File: assignments/hackyourtemperature/sources/keys.js [hunk 5]
```diff
@@ -0,0 +1,3 @@
+export const keys = {
```
Comment (EdwardSmart98) [inline assignments/hackyourtemperature/sources/keys.js:1]: It's good practice to not push up API keys to repo's to keep them safe. two ways to do this are:
-> Ignoring this file in the .gitignore.
-> adding a .env file (and ignoring it) - This is better, as the code will still compile, and the .env file is a standard across a lot of software/programs, so loading them into the program at runtime is handleded well.

**Review by EdwardSmart98 — APPROVED** [review]
A very strong assignment you should be proud of, some very nice features are showing through like separation of logic, and testing of edge cases. 
My biggest place for improvement is how can you better handle errors so that the front-end can use the messages in a valuable/consistent way.
